src/runtime/typemanager.cs

```diff
@@ -422,6 +434,198 @@ internal static IntPtr BasicSubType(string name, IntPtr base_, Type impl)
+            internal static int tp_str = 70;
+            internal static int tp_traverse = 71;
+            internal static int tp_members = 72;
+            internal static int tp_getset = 73;
+            internal static int tp_free = 74;
+            internal static int nb_matrix_multiply = 75;
+            internal static int nb_inplace_matrix_multiply = 76;
+            internal static int am_await = 77;
+            internal static int am_aiter = 78;
+            internal static int am_anext = 79;
+            internal static int tp_finalize = 80;
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        internal struct PY_TYPE_SLOT
+        {
+            public long slot; //slot id, from typeslots.h
+            public IntPtr func; //function pointer of the function implementing the slot
+        }
+
+        [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Ansi)]
+        internal class PyTypeSpecOffset
+        {
+            static PyTypeSpecOffset()
+            {
+                Type type = typeof(PyTypeSpecOffset);
+                FieldInfo[] fi = type.GetFields();
+                int size = IntPtr.Size;
+                for (int i = 0; i < fi.Length; i++)
+                {
+                    fi[i].SetValue(null, i * size + TypeOffset.ob_size);
+                }
+            }
+
+            public static IntPtr AllocPyTypeSpec(string typename, int obSize, int obFlags, IntPtr slotsPtr)
+            {
+                byte[] ascii = System.Text.Encoding.ASCII.GetBytes(typename);
+
+                //This approach is the same as the one in interop.cs for AllocModuleDef
+                //allocate the size of the struct (which is given by the value of the last
+                //static member and enough space to hold to typename as a char buffer.  The
+                //amount of space needed is the length of the string and the null terminator
+                //char* name member will simply point to the position of the buffer.
+                int size = name_value + ascii.Length + 1;
+                IntPtr ptr = Marshal.AllocHGlobal(size);
+
+                Marshal.Copy(ascii, 0, ptr + name_value, ascii.Length);
+                Marshal.WriteIntPtr(ptr, name, ptr + name_value);
+                Marshal.WriteByte(ptr, name + ascii.Length, 0);
+
+                Util.WriteCLong(ptr, basicsize, obSize);
+                Util.WriteCLong(ptr, itemsize, 0);
+                Util.WriteCLong(ptr, flags, obFlags);
+
+                Marshal.WriteIntPtr(ptr, slots, slotsPtr);
+                return ptr;
+            }
+
+            public static int name = 0;
+            public static int basicsize = 0;
+            public static int itemsize = 0;
+            public static int flags = 0;
+            public static int slots = 0;
+
+            private static int name_value = 0;
+        }
+
+
+        internal static IntPtr AllocateTypeObject(string name, PY_TYPE_SLOT[] type_slots)
```

Comment:
NIT: this does not just allocate. It actually creates the instance representing the type.